Add DirectCLR (#781)

KylevdLangemheen opened the pull request:
Issue #781 proposes adding DirectCLR to Lightly. In PR #963 a loss was proposed but never merged.

As the paper uses a version of [infoNCE that is identical to NTX-ent](https://discordapp.com/channels/@me/1399964647523876874/1400461672883159160), we can use our existing implementation of NTX-ent instead. This way, implementing DirectCLR becomes a simple matter of taking the first _d_ values of the embedding and applying the NTX-ent loss to that instead (if my understanding is correct :smile:  ).

As DirectCLR does away with a projection head, I did not see an obvious way to implement it under low level module blocks. The proposed implementation is added as a deprecated model, but perhaps it does not even need to get its own model file and can just live as an example.

This PR:
- Adds DirectCLR as deprecated model
- Adds example for DirectCLR
- Adds test for DirectCLR

This PR _does not_:
- Add DirectCLR to lightly cli
- Implement benchmarks for DirectCLR

Let me know if you would prefer to see a variant outside of the deprecated high-level models. I could make a "dummy" projection head which just slices the embedding for example.

KylevdLangemheen commented on examples/pytorch/directclr.py:
```diff
@@ -0,0 +1,56 @@
+from lightly.models import DirectCLR
+from lightly.transforms.simclr_transform import SimCLRTransform
+
+resnet = torchvision.models.resnet18()
```
Changed it, also for `torchvision.datasets`

KylevdLangemheen replied:
I don´t know exactly where the line is drawn (i.e. nn.Sequential). There's some variance in the codebase I think?

liopeer replied:
It's actually a hard line haha. :) Unfortunately a lot of the code is not up-to-date.

KylevdLangemheen commented on tests/models/test_ModelsDirectCLR.py:
```diff
@@ -0,0 +1,101 @@
+    return backbone
+
+
+class TestModelsDirectCLR(unittest.TestCase):
```
In the current set-up I've made DirectCLR loss inherit from the NTXent loss. As such, many tests become redundant. I currently test whether it can take various loss_dims, whether some of the calculations (that use NTXent) are done correctly, and whether the parent parameter (temperature) is set correctly. Let me know if some tests are missing or redundant still!

KylevdLangemheen commented on lightly/models/directclr.py:
```diff
@@ -0,0 +1,106 @@
+            return out0
+
+        # forward pass of second input x1
+        f1: Tensor = self.backbone(x1).flatten(start_dim=1)
```
Now that it is no longer a model but an example, I have included this with the example. Let me know if it is better to keep it simple in the example instead.

liopeer commented on lightly/loss/directclr_loss.py:
```diff
@@ -0,0 +1,91 @@
+    Attributes:
+        loss_dim:
+            Computes the loss only on the first loss_dim values of the encoding.
+        *args:
```
Since we have full control over the `NTXentLoss` and its arguments it's usually preferred to be verbose and copy the arguments from the NTXentLoss over here. I would only use `*args` and `**kwargs` when the subsequently instantiated class comes from a third-party package and the interface might change unexpectedly – and even in that case I would only allow passing keyword arguments and not positional arguments.

liopeer commented on examples/pytorch/directclr.py:
```diff
@@ -0,0 +1,53 @@
+from lightly.transforms.simclr_transform import SimCLRTransform
+
+resnet = models.resnet18()
+model = nn.Sequential(*list(resnet.children())[:-1])
```
Nit:
```suggestion
model = Sequential(*list(resnet.children())[:-1])
```

liopeer commented on lightly/loss/directclr_loss.py:
```diff
@@ -0,0 +1,91 @@
+    """Implementation of the NT-Xent based DirectCLR Loss.
+
+    Following the DirectCRL[0] paper, this loss should be used without projection
+    head. Set loss_dim to the desired truncated representation length.
```
Nit:
```suggestion
    head. Set `loss_dim` to the desired truncated representation length.
```
This will render nicely once we add it to the docs. Same thing further down.

liopeer commented on lightly/loss/directclr_loss.py:
```diff
@@ -0,0 +1,91 @@
+            DirectCLR Loss value.
+        """
+
+        out0 = out0.flatten(start_dim=1)[:, : self.loss_dim]
```
Flattening would not really be necessary since you have a really nice docstring that tells the user how it should be used. But you can leave it as is.

liopeer commented on tests/loss/test_directclr_loss.py:
```diff
@@ -0,0 +1,107 @@
+        dimension: int,
+        loss_dim: int,
+    ) -> None:
+        out0 = torch.tensor(
```
You could also directly use `torch.randn`, but it's of course totally okay this way as well.

liopeer commented on tests/loss/test_directclr_loss.py:
```diff
@@ -0,0 +1,107 @@
+        ), f"Expected temperature to be 0.5, but got {loss_function.temperature}"
+
+    @pytest.mark.parametrize("n_samples", [1, 2, 4])
+    @pytest.mark.parametrize("dimension", [1, 2, 16, 64])
```
I would use smaller dimensions here. We have a lot of unit tests in LightlySSL, so we should always try to use the fastest possible unit tests – ofc without compromising correctness.

liopeer commented on tests/loss/test_directclr_loss.py:
```diff
@@ -0,0 +1,107 @@
+
+    def test_forward_pass(self) -> None:
+        loss = DirectCLRLoss(loss_dim=32, temperature=0.5, memory_bank_size=0)
+        for bsz in range(1, 20):
```
Also here we should speed things up a notch. I would test max 2 cases: Batch size 1 and something else. And I would parametrize the test, so that you get a dedicated test result for each case. Parametrizing would also allow to use e.g. `pytest-xdist` to run the tests in parallel.

liopeer commented on tests/loss/test_directclr_loss.py:
```diff
@@ -0,0 +1,107 @@
+
+    def test_forward_pass_1d(self) -> None:
+        loss = DirectCLRLoss(loss_dim=32, temperature=0.5, memory_bank_size=0)
+        for bsz in range(1, 20):
```
Same thing here.

liopeer reviewed (changes requested):
Awesome work! It's looking really nice and many of my comments are really just nitpicking and beautification.

The two things we should really change however before merging: The `*args, **kwargs` story, and the *slightly too comprehensive* 🙃 unit tests. After that we're ready to merge.

yutong-xiang-97 commented:
Quick note: the `examples/` are for the supported models / SSL methods in the docs (like shown [here](https://docs.lightly.ai/self-supervised-learning/examples/simclr.html)) but not for the losses. We could have a follow-up issue implementing the whole DirectCLR at `benchmarks/`. Then we can add examples and docs for it.

yutong-xiang-97 reviewed (commented):
Added my review as well.

KylevdLangemheen commented:
I have:
- Generated the example notebooks with `make generate-example-notebooks`
- Reduced dims in tests
- Fixed typo
- Added formatting for loss_dim
- Directly import functions/classes at the module level
- Use `torch.randn` consistently
- Added NTXentLoss arguments directly

Is there a standard way to ensure or even automate that the parameters relating to NTXentLoss always match? This way the defaults can get out of sync pretty easily :(

KylevdLangemheen commented:
Resolved the temperature parameter and default parameter comments.

liopeer reviewed (approved):
Everything addressed from my side as well. Amazing work @KylevdLangemheen !
